File: src/main/java/shopping/ui/ProductController.java
package shopping.ui;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.servlet.support.ServletUriComponentsBuilder;
import shopping.application.product.ProductService;
import shopping.dto.ProductRequest;
import shopping.dto.ProductResponse;

import java.net.URI;
import java.util.List;

@RestController
@RequestMapping("/api/products")
public class ProductController {
    private final ProductService productService;

    public ProductController(ProductService productService) {
        this.productService = productService;
    }

    @PostMapping
    public ResponseEntity<Void> create(@RequestBody ProductRequest request) {
        Long productId = productService.createProduct(request);

        URI location = ServletUriComponentsBuilder.fromCurrentRequest()
                .path("/{id}")
                .buildAndExpand(productId)
                .toUri();
        return ResponseEntity.created(location).build();
    }

    @GetMapping("/{productId}")
    public ResponseEntity<ProductResponse> getProduct(@PathVariable Long productId) {
        ProductResponse response = productService.getProduct(productId);
        return ResponseEntity.ok(response);
    }

    @GetMapping
    public ResponseEntity<List<ProductResponse>> getProducts() {
        return ResponseEntity.ok(productService.getAllProducts());
    }

    @PutMapping("/{productId}")
    public ResponseEntity<Void> updateProduct(
            @PathVariable Long productId,
            @RequestBody ProductRequest request) {
        productService.updateProduct(productId, request);
        return ResponseEntity.ok().build();
    }

    @DeleteMapping("/{productId}")
    public ResponseEntity<Void> deleteProduct(@PathVariable Long productId) {
        productService.deleteProduct(productId);
        return ResponseEntity.noContent().build();
    }
}

File: src/main/kotlin/shopping/interfaces/api/product/ProductController.kt
package shopping.interfaces.api.product

import com.failsafe.interfaces.api.ApiResponse
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import shopping.application.product.ProductFacade
import shopping.interfaces.api.Paging

@RestController
@RequestMapping("/api/products")
class ProductController(
    private val productFacade: ProductFacade,
) {
    @GetMapping
    fun getProducts(
        @RequestBody request: ProductDto.ProductSearchRequest,
    ): ApiResponse<Paging.PageResponse<ProductDto.ViewResponse>> =
        productFacade
            .getProducts(request.paging)
            .map { ProductDto.ViewResponse.from(it) }
            .let { Paging.PageResponse.from(it) }
            .let { ApiResponse.success(it) }

    @PostMapping
    fun createProduct(
        @RequestBody request: ProductDto.CreateProductRequest,
    ): ApiResponse<ProductDto.ViewResponse> =
        productFacade
            .createProduct(request)
            .let { ProductDto.ViewResponse.from(it) }
            .let { ApiResponse.success(it) }

    @GetMapping("/{productId}")
    fun getProduct(
        @PathVariable productId: Long,
    ): ApiResponse<ProductDto.ViewResponse> =
        productFacade
            .getProduct(productId)
            .let { ProductDto.ViewResponse.from(it) }
            .let { ApiResponse.success(it) }

    @PutMapping("/{productId}")
    fun updateProduct(
        @PathVariable productId: Long,
        @RequestBody request: ProductDto.UpdateProductRequest,
    ): ApiResponse<ProductDto.ViewResponse> =
        productFacade
            .updateProduct(productId, request)
            .let { ProductDto.ViewResponse.from(it) }
            .let { ApiResponse.success(it) }

    @DeleteMapping("/{productId}")
    fun deleteProduct(
        @PathVariable productId: Long,
    ): ApiResponse<Any> {
        productFacade.deleteProduct(productId)
        return ApiResponse.success()
    }
}
